Add load balancer and IP pool docs (#323)
harvester/docs: docs/networking/ippool.md

---
sidebar_position: 5
sidebar_label: IP Pool
title: "IP Pool"
keywords:
- IP Pool
---
_Available as of v1.2.0_

> Harvester IP Pool is a built-in IP address management (IPAM) solution exclusively available to Harvester Load Balancers (LBs).

Reviewer: ```suggestion
Harvester IP Pool is a built-in IP address management (IPAM) solution exclusively available to Harvester load balancers (LBs).
```

### Examples
- **Example 1:** You want to configure an IP pool with the range `192.168.100.0/24` for the namespace `default`. All the load balancers in the namespace `default` get an IP from this IP pool. For an example, refer to the following `.yaml`:
  
  ```yaml
  apiVersion: networking.harvesterhci.io/v1beta1
  kind: IPPool
  metadata:
    name: default-ip-pool
  spec:
    ranges:
    - subnet: 192.168.100.0/24
    selector:
      scope:
        namespace: default
  ```

- **Example 2:** You have a guest cluster `rke2` deployed with network `default/vlan1` in the project/namespace `product/default` and want to configure an exclusive IP pool with range `192.168.10.10-192.168.10.20` for it. For an example, refer to the following `.yaml`:
  
  ```yaml
  apiVersion: networking.harvesterhci.io/v1beta1
  kind: IPPool
  metadata:
    name: rke2-ip-pool
  spec:
    ranges:
    - subnet: 192.168.10.0/24
      rangeStart: 192.168.10.10
      rangeEnd: 192.168.10.20
    selector:
      network: default/vlan1
      scope:
        project: product
        namespace: default
        cluster: rke2
  ```

> - **Example 3:** You want to migrate the IP pool `default-ip-pool` to a different IP pool `default-ip-pool-2` with range `192.168.200.0/24`. The IP pool `default-ip-pool` has a higher priority than `default-ip-pool`. For an example, refer to the following `.yaml`:

Author: The IP pool `default-ip-pool-2` has a higher priority than `default-ip-pool`.

  
  ```yaml
  apiVersion: networking.harvesterhci.io/v1beta1
  kind: IPPool
  metadata:
    name: default-ip-pool-2
  spec:
    ranges:
    - subnet: 192.168.200.0/24
    selector:
      priority: 1
      scope:
        namespace: default
  ```

> - Example 4: You want to configure a global IP pool with range `192.168.20.0/24`. For an example, refer to the following `.yaml`:

Reviewer: ```suggestion
- **Example 4:** You want to configure a global IP pool with range `192.168.20.0/24`. For an example, refer to the following `.yaml`:
```

  
  ```yaml
  apiVersion: networking.harvesterhci.io/v1beta1
  kind: IPPool
  metadata:
    name: global-ip-pool
  spec:
    ranges:
    - subnet: 192.168.20.0/24
    selector:
      scope:
        project: "*"
        namespace: "*"
        cluster: "*"
  ```

## Allocation policy
- The IP pool prefers to allocate the previously assigned IP address according to the given history.
- IP address allocation follows the round-robin policy.

## How to create
To create a new VM load balancer:

> 1. Go to the **Networks > IP Pools** page and select **Create**.

Reviewer: ```suggestion
1. Go to the **Networks** > **IP Pools** page and select **Create**.
```

1. Specify the **Name** of the IP pool.
1. Go to the **Range** tab to specify the **IP ranges** for the IP pool. You can add multiple IP ranges.
1. Go to the **Selector** tab to specify the **Scope** and **Priority** of the IP pool.

:::note

> Starting with Harvester v1.2.0, vip-pools setting is deprecated. After upgrading, vip-pools setting will be automatically converted to IP pools.

Reviewer: ```suggestion
Starting with Harvester v1.2.0,  the `vip-pools` setting is deprecated. After upgrading, the `vip-pools` setting will be automatically converted to IP pools.
```

## Features
- Multiple IP ranges: Each IP pool can contain multiple IP ranges or CIDRs.
- Allocation history: The IP pool keeps track of the allocation history of every IP address and prioritizes assigning addresses that were previously allocated.
- Scope: The IP pool can be restricted to a particular network, project, namespace, or guest cluster.

## Selection Policy
As the IP pool is designated for specific usage scopes, you can specify the corresponding requirements in the LB, and the IP pool that meets the requirements will automatically assign IP addresses for the LB.

- The LB expresses the requirements by the annotations outlined below. All the annotations are optional.
  - `loadbalancer.harvesterhci.io/network` specifies the location of the network for the guest cluster.
  - `loadbalancer.harvesterhci.io/project`, `loadbalancer.harvesterhci.io/namespace` specify the project and namespace of the VMs that comprise the guest cluster.
  - `loadbalancer.harvesterhci.io/cluster` specifies the name of the guest cluster.
- The IP pool has a selector, including network and scope, to match the requirements of the LB.

>   - Network is a hard condition. The optional IP pool must match the value of the LB annotation `loadbalancer.harvesterhci.io/network`.

Author: No, it's different.

---
sidebar_position: 5
sidebar_label: IP Pool
title: "IP Pool"
keywords:
- IP Pool
---
_Available as of v1.2.0_

> Harvester IP Pool is a built-in IP address management (IPAM) solution that is exclusively available to Harvester Load Balancers (LBs).

Reviewer: ```suggestion
Harvester IP Pool is a built-in IP address management (IPAM) solution exclusively available to Harvester Load Balancers (LBs).
```

## Features

> - Multiple IP ranges: Each IP pool can contain multiple IP ranges or CIDRs.
> - Allocation history: The IP pool keeps track of the allocation history of every IP address and prioritizes assigning addresses that were previously allocated.
> - Scope: The IP pool can be restricted to a particular network, project, namespace, or guest cluster.

Reviewer: ```suggestion
- **Multiple IP ranges:** Each IP pool can contain multiple IP ranges or CIDRs.
- **Allocation history:** The IP pool keeps track of the allocation history of every IP address and prioritizes assigning previously allocated addresses.
- **Scope:** Restrict the IP pool to a particular network, project, namespace, or guest cluster.
```

## Selection policy

> As the IP pool is designated for specific scopes, you can specify the corresponding requirements in the LB, and the IP pool that meets the requirements will automatically assign IP addresses for the LB.
> 
> - The LB uses the annotations below to express the requirements. All the annotations are optional.
>   - `loadbalancer.harvesterhci.io/network` specifies the location of the network for the guest cluster.
>   - `loadbalancer.harvesterhci.io/project` and `loadbalancer.harvesterhci.io/namespace` specify the project and namespace of the VMs that comprise the guest cluster respectively.
>   - `loadbalancer.harvesterhci.io/cluster` specifies the name of the guest cluster.
> - The IP pool has a selector, including network and scope, to match the requirements of the LB.
>   - Network is a hard condition. The optional IP pool must match the value of the LB annotation `loadbalancer.harvesterhci.io/network`.
>   - Every IP pool, except the global IP pool, has a unique scope different from others if its priority is `0`. The project/namespace/cluster name of LBs should be in the scope of the IP pool if they want to get an IP from this pool.
> - `spec.selector.priority` specifies the priority of the IP Pool. The larger the number, the higher the priority. If the priority is not 0, its value should be different from others. The priority helps you to migrate the old IP pool to the new one.
> - If the IP Pool has a scope that matches all projects, namespaces, and guest clusters, it's called a global IP pool. It's only allowed to have one global IP pool. If there is no IP pool matching the requirements of the LB, the IPAM will allocate an IP from the global IP pool if it exists.

Reviewer: ```suggestion
An IP pool can have specific scopes, and you can specify the corresponding requirements in the LB. The IP pool that meets the requirements will automatically assign IP addresses for the LB.

- The LB uses the annotations below to express the requirements, and all the annotations are optional.
  - `loadbalancer.harvesterhci.io/network` specifies the location of the network for the guest cluster.
  - `loadbalancer.harvesterhci.io/project` and `loadbalancer.harvesterhci.io/namespace` specify the project and namespace of the VMs that comprise the guest cluster.
  - `loadbalancer.harvesterhci.io/cluster` specifies the name of the guest cluster.
- The IP pool has a selector, including network and scope, to match the requirements of the LB.
  - Network is a hard condition. The optional IP pool must match the value of the LB annotation `loadbalancer.harvesterhci.io/network`.
  - Every IP pool, except the global IP pool, has a unique scope different from others if its priority is `0`. The project, namespace, or cluster name of LBs should be in the scope of the IP pool if they want to get an IP from this pool.
- `spec.selector.priority` specifies the priority of the IP Pool. The larger the number, the higher the priority. If the priority is not `0`, the value should differ. The priority helps you to migrate the old IP pool to the new one.
- If the IP Pool has a scope that matches all projects, namespaces, and guest clusters, it's called a global IP pool. It's only allowed to have one global IP pool. If there is no IP pool matching the requirements of the LB, the IPAM will allocate an IP from the global IP pool if it exists.
```

## Allocation policy
- The IP pool prefers to allocate the previously assigned IP according to the given history.
- IP allocation follows the round-robin policy.

## How to create

> To create a new VM load balancer:
> 1. Go to the **Networks > IP Pools** page and click the **Create** button.
> 1. Specify the name of the IP pool.
> 1. Go to the **Range** tab to specify the IP ranges for the IP pool. You can add multiple IP ranges.
> 1. Go to the **Selector** tab to specify the scope and priority of the IP pool.

Reviewer: ```suggestion
To create a new VM load balancer:

1. Go to the **Networks > IP Pools** page and select **Create**.
1. Specify the **Name** of the IP pool.
1. Go to the **Range** tab to specify the **IP ranges** for the IP pool. You can add multiple IP ranges.
1. Go to the **Selector** tab to specify the **Scope** and **Priority** of the IP pool.
```
